### ISSUE TITLE
A faster (tested only on V100) GEMM kernel dedicated for small batch-size

### ISSUE THREAD

**Issue by ZiyueHuang** [vllm-project/vllm#2300]
Inspired by https://github.com/wangsiping97/FastGEMV, I took some time to implement a custom GEMM kernel. Compared to FastGEMV, it does not use shared memory and supports batch size larger than 1.
- For small batch size (in the decoding stage), it seems to be faster than `F.linear` currently used in vllm, and should be easy to fuse with the custom allreduce kernels here (https://github.com/vllm-project/vllm/pull/2192).
- Only tested on V100, torch=='2.0.1+cu117'. Not sure about the performance on other GPUs or other CUDA versions.

Here is implementation (https://github.com/ZiyueHuang/vllm/commit/20e83b0a8b7518eb905cdc8b99fba32fc7d7c96d). Below is the reproducible script for performance benchmarks (including the correctness checking at the warmup stage).
<details>
<summary>Click here for the script</summary>

```
import torch
import math
import time
from vllm import custom_gemm

import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--use_custom', action='store_true')
parser.add_argument('--bs', type=int)
parser.add_argument('--ic', type=int)
parser.add_argument('--oc', type=int)
args = parser.parse_args()

bs = args.bs
ic = args.ic
oc = args.oc

warmup = 5
nrepeats = 16

inputs_l = [5.*math.sqrt(ic)/ic * torch.randn((bs, ic), dtype=torch.half, device=torch.device('cuda')) \
                for _ in range(nrepeats)]
weight_l = [5.*math.sqrt(ic)/ic * torch.randn((oc, ic), dtype=torch.half, device=torch.device('cuda')) \
                for _ in range(nrepeats)]

for _ in range(warmup):
    out = torch.nn.functional.linear(inputs_l[0], weight_l[0])
    out2 = custom_gemm.gemm_small_bs(inputs_l[0], weight_l[0])
    #print(out[0, 50:60])
    #print(out2[0, 50:60])
    torch.testing.assert_close(out, out2, rtol=1e-3, atol=5e-4)

torch.cuda.synchronize()

tik = time.time()

if args.use_custom:
    for i in range(len(inputs_l)):
        out = custom_gemm.gemm_small_bs(inputs_l[i], weight_l[i])
else:
    for i in range(len(inputs_l)):
        out = torch.nn.functional.linear(inputs_l[i], weight_l[i])

torch.cuda.synchronize()
print(str((bs, ic, oc)) + "##" + str(time.time() - tik))
```
</details>

Performance report for different shapes of GEMM used in Qwen-7B, Qwen-14B, Qwen-14B (with tensor-parallel-size=2), where `oc` is the number of output channels and `ic` is the number of input channels, and the numbers are `time(custom-gemm) /  time(F.linear)` (numbers smaller than 1 implies better performance of custom-gemm):

| batch_size / (oc, ic) | (22016, 4096) | (4096, 11008) | (12288, 4096) | (4096, 4096) |
| :----------------:    | :------:      | :----:        | :----:        | :----:       |
| 1    |    0.783    |    0.899    |   0.823   |  0.817 |
| 2    |    0.784    |    0.658    |   0.802   |  0.783 |
| 3    |    0.791    |    0.624    |   0.804   |  0.777 |
| 4    |    0.840    |    0.707    |   0.831   |  0.837 |

| batch_size / (oc, ic) | (27392, 5120) | (5120, 13696) | (15360, 5120) | (5120, 5120) |
| :----------------:    | :------:      | :----:        | :----:        | :----:       |
| 1    |    0.744    |    0.865    |   0.596   |  0.827 |
| 2    |    0.756    |    0.670    |   0.605   |  0.624 |
| 3    |    0.765    |    0.716    |   0.618   |  0.678 |
| 4    |    0.791    |    0.818    |   0.634   |  0.755 |

| batch_size / (oc, ic) | (13696, 5120) | (5120, 6848) | (7680, 5120) | (5120, 2560) |
| :----------------:    | :------:      | :----:        | :----:        | :----:       |
| 1    |    0.732    |    1.058    |   0.832   |  0.839 |
| 2    |    0.740    |    0.878    |   0.723   |  0.704 |
| 3    |    0.761    |    0.973    |   0.742   |  0.735 |
| 4    |    0.813    |    1.216    |   0.789   |  0.792 |

**Comment by hanzhi713** [comment]
Did you capture profile with nsys? Launch overhead (torch overhead, GEMM heuristics and actually launching the kernel) for small matmuls might be significant, and I'm afraid measuring like this might be inaccurate.

**Comment by ZiyueHuang** [comment]
Hmm, you are right. I forgot to exclude the overhead. The total execution time is around hundreds of us to a few ms, and the diff is even smaller. Will do more profile. Thanks!

**Comment by ZiyueHuang** [comment]
@hanzhi713 I did the profile by nsys, and update the numbers (`time(custom-gemm) / time(F.linear)`) below. 

I only finish the numbers up to batch-size=2 at the moment, because I don't know how to write a script to automatically generate all these profiling numbers from nsys and it is laborious to compute these numbers one-by-one.

Qwen-7B:
| batch_size \ (oc, ic) | (22016, 4096) | (4096, 11008) | (12288, 4096) | (4096, 4096) |
| :----------------:    | :------:      | :----:        | :----:        | :----:       |
| 1    |    0.792    |    0.914    |   0.831   |  0.924 |
| 2    |    0.788    |    0.670    |   0.796   |  0.805 |


Qwen-14B:
| batch_size \ (oc, ic) | (27392, 5120) | (5120, 13696) | (15360, 5120) | (5120, 5120) |
| :----------------:    | :------:      | :----:        | :----:        | :----:       |
| 1    |    0.751    |    0.947    |   0.599   |  0.857 |
| 2    |    0.762    |    0.751    |   0.610   |  0.650 |

Qwen-14B with tensor-parallel-size=2
| batch_size \ (oc, ic) | (13696, 5120) | (5120, 6848) | (7680, 5120) | (5120, 2560) |
| :----------------:    | :------:      | :----:        | :----:        | :----:       |
| 1    |    0.740    |    1.08    |   0.840   |  0.876 |
| 2    |    0.751    |    0.902    |   0.731   |  0.733 |
